### ISSUE TITLE
[feature request] [pytorch] Truncated SVD

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#8049]
Facebook published reference code of fast NumPy randomized SVD (based on power iteration): https://github.com/facebook/fbpca/blob/master/fbpca.py#L1503

Only top eigenvectors are useful e.g. for spectral clustering, it would be nice to have it supported for CUDA.

**Comment by SsnL** [comment]
I only skimmed through the code. It seems very doable in PyTorch. Have you tried directly translating into PyTorch?

**Comment by vadimkantorov** [comment]
In the blog post they explain: "to maintain numerical precision throughout these operations, we take intermediate QR and LU decompositions during the power iterations."

I skimmed PyTorch docs, it does have `potrf` for p.s.d matrices, but not a general LU-decomp?

But it's surely possible to do without LU-decomp, and just do a stock SVD on a smaller matrix, like is done in many tutorials.

**Comment by vadimkantorov** [comment]
It would also be nice to benchmark it against https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html for symmetric sparse matrices (coming from k-NN approximations to graph laplacians)

**Comment by vadimkantorov** [comment]
As @karol-arndt pointed out, LU-decomp exists as `torch.btrifact` - and this should be the last element for porting the randomized-svd code.

**Comment by vadimkantorov** [comment]
two existing implementations in sklearn that may be useful for API design:
http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.TruncatedSVD.html
http://scikit-learn.org/dev/modules/generated/sklearn.utils.extmath.randomized_svd.html

**Comment by vadimkantorov** [comment]
@vishwakftw if you are on a linear algrebra impl streak, you may like this one :)

**Comment by lobpcg** [comment]
LOBPCG is expected to outperform randomized SVD (based on power iteration), see
https://github.com/scikit-learn/scikit-learn/pull/12319

LOBPCG is already available in CUDA and MAGMA, so may be not so difficult to add it to PyTorch.

**Comment by Jbollenbacher** [comment]
Any updates on a native pytorch implementation for this?  It would be very useful for training svd-compressed deep networks.

**Comment by lobpcg** [comment]
Such LOBPCG implementation is still at the planning stage; see https://github.com/lobpcg/blopex-python/blob/master/README.md

**Comment by zhangguanheng66** [comment]
Thanks @lobpcg . I will take a look and implement it soon.

**Comment by lobpcg** [comment]
> Thanks @lobpcg . I will take a look and implement it soon.

Great,m thanks! I am very busy and expect it getting worse, so I would not be able to do any coding myself, but glad to help with advice.  Please let me know if you have any questions on the abstract implementation proposed in  https://github.com/lobpcg/blopex-python/blob/master/README.md Do you want me to add you to this repository?

**Comment by zhangguanheng66** [comment]
That would be great.@lobpcg

**Comment by lobpcg** [comment]
Done, please accept. 

> That would be great.@lobpcg

**Comment by ezyang** [comment]
@Roger-luo, when he came and visited Facebook, said this would be very useful for tensor networks!

**Comment by gokceneraslan** [comment]
`fbpca` also supports PCA with sparse matrices without centering and densifying them. There is also a nice discussion and PR on sklearn github about PCA with sparse matrices: https://github.com/scikit-learn/scikit-learn/issues/12794

It'd be super cool to have this in pytorch, too.

**Comment by rumschuettel** [comment]
> Thanks @lobpcg . I will take a look and implement it soon.
@zhangguanheng66 did you have a chance to have a look at this? It would be extremely interesting to have it supported in torch directly, instead of taking the full svd and tossing away almost all of the vectors.

Thanks so much!

**Comment by zhangguanheng66** [comment]
@rumschuettel I'm currently busy with other stuffs. If anyone are interested in the task, please feel free to take it.

**Comment by vadimkantorov** [comment]
@pearu After your PR is merged, perhaps, we can keep this issue open for other points discussed here: 1) sparse PCA, 2) LOBPCG algorithm (PR in SciPy seems progressing as well)

**Comment by pearu** [comment]
@vadimkantorov , I planned to address the sparse/batched SVD/PCA, as well as LOBPCG in the PR #29488  . If it is going to be too large a task, perhaps we should create a separate issue for each point.
But I agree, all points discussed in this issue should be addressed before closing it.

**Comment by vadimkantorov** [comment]
awesome stuff :) are these implementations differentiable wrt input?

**Comment by vadimkantorov** [comment]
@pearu @nikitaved For many cases it's enough to have just the largest (smallest) eigenvalue / eigenvector (originally asked in #12760) - e.g. for binary spectral clustering or for spectral normalization. In the simplest case this can be done with simple power iteration (as already done in the impl of spectral_norm). Power iteration can also be differentiated through out of the box, especially if number of iterations is not very large.

@pearu Will the new svd_lowrank work well for this common scenario out-of-the-box? If yes, then I guess describing this clearly in the docs is enough (along with the algorithm used for the task). If yes, then the custom spectral_norm impl can be eliminated - it can be a good usage example of these new APIs. 

#12760 also proposes exposing the largest eigenvalue matrix norm from the torch.norm interface

**Comment by vadimkantorov** [comment]
@pearu Do I understand correctly that the Halko et al paper does also some kind of (more advanced) randomized svd similar to fbpca?

**Comment by vadimkantorov** [comment]
@pearu @nikitaved What's your take on https://github.com/pytorch/pytorch/issues/8049#issuecomment-588188616 ?

**Comment by nikitaved** [comment]
@vadimkantorov, let's probably write that proposal discussed in [#29488](https://github.com/pytorch/pytorch/pull/29488) where we could keep a list of desirable algorithms?
You can kick off with the largest eigenvector. I also have some things in my mind.
Will you be able to formulate it? I will join you in once I have a bit of free time.

**Comment by pearu** [comment]
@vadimkantorov sorry for the late response.
The algorithm of `svd_lowrank` assumes that it's input is a lowrank matrix. Applying `svd_lowrank` to a full-rank matrix violates this assumption. So, I would not trust that the result will give extreme singular values exactly, although, it may be close to it. This requires experimenting (that may lead to discarding the idea) and analysis (to ensure that the method gives always expected results).

**Comment by vadimkantorov** [comment]
@nikitaved My feature request is some recommendations how to retrieve top eigenvalue/vector from a matrix (can be low-rank or full-rank, full or sparse). If it's already satisfied - then just documentation improvements will do.

Currently this is done in spectral normalization with a simple power iteration. If it's a more efficient method than svd_lowrank/lobpcg, then I propose to also surface a simple (randomized?) power iteration to the user. If svd_lowrank/lobpcg is more efficient, I propose to rewrite spectral normalization in terms of new functions and mention it as an example in svd_lowrank/lobpcg docs.

**Comment by nikitaved** [comment]
@vadimkantorov, no, there is no guarantee that the implemented method is "better" than the power method. I also think it would be great to have these methods in the randomized manifestation because it is very cheap. Do you really need high quality approximations for the spectral normalization in DNNs?

**Comment by vadimkantorov** [comment]
I filed this issue when I was working with various graph Laplacians. Some of them hard sharp-dropping spectrum, some had more linear dropping spectrum. When Laplacian was sharp-dropping, even a few iterations of power iteration sufficed. When spectrum didn't have a good gap, I had to use something more complex. So back in the day I filed this issue.

In PyTorch currently spectral norm is done in https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L37

Yes, if it's simple, a (optionally randomized) power iteration backend would be nice to have as well. Maybe @lobpcg has some advice? From previous messages I understood that lobpcg is almost always "better" than randomized power iteration. If it's the case, maybe we just need to use LOBPCG everywhere even for spectral normalization :)

**Comment by nikitaved** [comment]
Ah, so you work with PSD only? If so, then the current randomized implementation is not tuned for PSDs. It has to be extended to treat PSDs better.

**Comment by nikitaved** [comment]
Well, "better" is rather vague. You have to decide the performance-accuracy trade off. I know that spectral normalization has recently gained some attention. I will read the paper to see whether the approach is actually sensitive to accuracy.
Of course, if you perform something like Spectral Clustering, I would assume you need accuracy over performance for interoperability reasons...

**Comment by vadimkantorov** [comment]
(in my particular usecase it was spectral clustering, but done within a neural net training iteration, so I preferred performance over accuracy. Laplacian was PSD, my graph size was quite small, and power iteration was fast/accurate enough)

**Comment by lobpcg** [comment]
1. simple (randomized) power iteration is simple to code, numerically stable, and has low cost per iteration. 
 2. LOBPCG is more sophisticated to go around numerical instabilities, i.e. harder to code, the cost per iteration is higher (probably in the worst case up to 50% larger), but MUCH faster convergence. 
 3. For "small" dense matrices, or when a "large" portion of eigenvectors is needed there is also a potentially faster and much simpler option just to compute ALL eigenvalues and eigenvectors, calling a library function.  Notice that the scipy LOBPCG implementation automatically switches to a standard dense solver internally if over 20% of the  eigenvectors is requested, taking advantage of this approach.

I have not actually seen or published myself a systematic numerical comparison. The above is based on a common knowledge of the algorithms and my limited testing.  

Having LOBPCG already implemented, I see no point to bother implementing in addition the simple (randomized) power iteration, as the potential gain, if any, is unlikely significant to justify writing and maintaining another code . 

In the opposite direction, having randomized power iteration already implemented, an effort is still needed to convince the package maintainers to include LOBPCG, e.g., see the stalled https://github.com/scikit-learn/scikit-learn/pull/12319 awaiting good and convincing numerical comparisons.

**Comment by lobpcg** [comment]
One should be easily able to use this truncated SVD with LOBPCG solver as an alternative to  PyTorch current spectral norm calculator by the power/simple iteration in https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L37

Indeed, may be an example like this should be added to the documentation at least.

If any reasonable accuracy is needed, SVD via LOBPCG should very likely outperform the power/simple for spectral normalization, especially for unevenly sized matrices, i.e.  really tall or fat.

**Comment by nikitaved** [comment]
@vadimkantorov , you mean this paper, right? https://arxiv.org/pdf/1802.05957.pdf

**Comment by vadimkantorov** [comment]
@nikitaved Yes, as mentioned in https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/spectral_norm.py#L2

**Comment by nikitaved** [comment]
OK, then it looks like they are working on estimating the Lipschitzness of the objective function. Any overestimation will do, so I am not sure that high accuracy is needed. I guess you are right, it makes sense to give a chance to choose between the methods. I have no idea how it will affect the performance in practice. It seems like in the worst case the training time might get affected, not the overall performance. Definitely requires some testing...

**Comment by lobpcg** [comment]
Both the power iteration and LOBPCG underestimate the spectral norm. A correction is needed to compute a guaranteed overestimate. However, in this paper iterations of the power method are combined with step (23), which I don't understand, so this is not just a spectral normalization. 
null

**Comment by nikitaved** [comment]
It is a single step of Stochastic Gradient Descent. As you can see they perform this normalization at each step of SGD for each weight matrix. The speed could of very high performance in here...

**Comment by lobpcg** [comment]
Thanks! I missed the header. So, it's SGD with extra normalization steps. Makes sense, except that I would initiate the SDG iterations with already fully normalized W. Yes, a few iterations of SVD based on LOBPCG could substitute the single power iteration here. Easy to try. May help if W changes noticeable between the SGD iterations. 
null

**Comment by nikitaved** [comment]
@lobpcg, is it cheap to compute that correction in order to get an overestimation?

**Comment by lobpcg** [comment]
@nikitaved - basically, yes.

In LOBPCG, the overestimation is already available directly from the code output in this case.
 
It should be possible to compute the overestimation for that power method, too, but it first requires deriving the formula for computation from the vectors available , and I have not looked at it.

**Comment by nikitaved** [comment]
I can give it a try. I will have a look how LOBPCG does that...

**Comment by lobpcg** [comment]
@nikitaved - great! 

May be first just put LOBPCG there replacing the power method and limit the number of LOBPCG iterations to 2 or 3. LOBPCG can also reuse the approximation to the singular vector from the previous SGD iteration. In contrast to that power method, LOBPCG iterates only one set of singular vectors, left or right, whichever is smaller size. Initiate LOBPCG with a single random vector outside of SGD loop. The costs/outcome should be comparable to that after the power method. 

You may want to print the obtained scaling factor in both approaches. Both power and LOBPCG underestimate, but LOBPCG value should be larger on the same W, i.e., better.

Please feel free to ping me after this implementation is done, and I show you how to compute the overestimation from the LOBPCG outputs.
